cdap-ranger/cdap-ranger-lookup/src/main/java/co/cask/cdap/security/authorization/ranger/lookup/client/CDAPClient.java

```diff
@@ -0,0 +1,226 @@
+  private final String password;
+  private final NamespaceClient nsClient;
+  private static final String ERR_MSG = " You can still save the repository and start creating "
+    + "policies, but you would not be able to use autocomplete for "
+    + "resource names. Check xa_portal.log for more info.";
+
+  CDAPClient(String serviceName, String instanceURL, String username, String password) {
+    this.serviceName = serviceName;
+    this.instanceURL = instanceURL;
+    this.username = username;
+    this.password = password;
+    initConnection();
+    ClientConfig clientConfig = getClientConfig();
+    this.nsClient = new NamespaceClient(clientConfig);
+    //TODO create more cdap clients here
+  }
+
+  private void initConnection() {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> CDAPClient initConnection()");
+    }
+    if (username != null && password != null) {
+      // security is enabled, we need to get access token before checking system services
+      try {
+        LOG.info(String.format("Fetching access token with username %s and password ****", username));
+        accessToken = fetchAccessToken(username, password);
+      } catch (Exception ex) {
+        throw Throwables.propagate(ex);
+      }
+    }
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("<== CDAPClient initConnection()");
+    }
+  }
+
+  /**
+   * Tests that a connection can be established to CDAP with the given config. We do this by listing namespaces.
+   * @return Map<String, Object> Connection test response
+   */
+  Map<String, Object> testConnection() {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> CDAPClient testConnection()");
+    }
+    HashMap<String, Object> responseData = new HashMap<>();
+    List<String> testResult;
+    try {
+      testResult = getNamespaces(null);
+      if (testResult != null && testResult.size() >= 0) {
+        String successMsg = "Connection Test Successful";
+        BaseClient.generateResponseDataMap(true, successMsg, successMsg,
+                                           null, null, responseData);
+      } else {
+        String failureMsg = "Unable to retrieve any namespace using given parameters. " +
+          "Please ensure that the configurations are correct and at least one namespace is accessible to this user.";
+        BaseClient.generateResponseDataMap(false, failureMsg, failureMsg + ERR_MSG,
+                                           null, null, responseData);
+      }
+    } catch (Exception e) {
+      LOG.error("Error connecting to CDAP.", e);
+      String failureMsg = "Unable to connect to CDAP instance." + e.getMessage();
+      BaseClient.generateResponseDataMap(false, failureMsg,
+                                         failureMsg + ERR_MSG, null, null, responseData);
+    }
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("<== CDAPClient testConnection()");
+    }
+    return responseData;
+  }
+
+  private List<String> getNamespaces(List<String> nsList) throws Exception {
```

Comment:
What is the usage of this method? The only place it is used is to get the namespaces for test connection. If that is the case why do we need to pass a list here, and we can just return `nsClient.list()`?